core/store/src/trie/mem/resharding.rs

```diff
@@ -0,0 +1,347 @@
+use crate::trie::{MemTrieChanges, TrieRefcountDeltaMap};
+use crate::{NibbleSlice, TrieChanges};
+
+use super::flexible_data::children::ChildrenView;
+use super::flexible_data::value::ValueView;
+use super::node::{MemTrieNodeId, MemTrieNodeView};
+use super::updating::{
+    MemTrieUpdate, OldOrUpdatedNodeId, UpdatedMemTrieNode, UpdatedMemTrieNodeId,
+};
+use super::{arena::ArenaMemory, node::MemTrieNodePtr};
+use itertools::Itertools;
+use near_primitives::challenge::PartialState;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::AccountId;
+use std::collections::HashMap;
+use std::ops::Range;
+use std::sync::Arc;
+
+/// Whether to retain left or right part of trie after shard split.
+pub enum RetainMode {
+    Left,
+    Right,
+}
+
+/// Decision on the subtree exploration.
+#[derive(Debug)]
+enum RetainDecision {
+    /// Retain the whole subtree.
+    RetainAll,
+    /// The whole subtree is not retained.
+    NoRetain,
+    /// Descend into all child subtrees.
+    Descend,
+}
+
+/// Tracks changes to the trie caused by the retain.
+struct UpdatesTracker {
+    /// Accessed node hashes and their serializations. Used for proof
+    /// generation.
+    #[allow(unused)]
+    node_accesses: HashMap<CryptoHash, Arc<[u8]>>,
+    /// All new nodes to be created.
+    updated_nodes: Vec<UpdatedMemTrieNode>,
+    /// On-disk changes applied to reference counts of nodes. In fact, these
+    /// are only increments.
+    refcount_changes: TrieRefcountDeltaMap,
+}
+
+impl UpdatesTracker {
+    pub fn new() -> Self {
+        Self {
+            node_accesses: HashMap::new(),
+            updated_nodes: Vec::new(),
+            refcount_changes: TrieRefcountDeltaMap::new(),
+        }
+    }
+
+    pub fn add_node(&mut self, node: UpdatedMemTrieNode) -> UpdatedMemTrieNodeId {
+        let id = self.updated_nodes.len();
+        debug_assert!(node != UpdatedMemTrieNode::Empty);
+        self.updated_nodes.push(node);
+        // TODO(#12074): apply remaining changes to `updates_tracker`.
+
+        id
+    }
+}
+
+impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
+    /// Splits the trie, separating entries by the boundary account.
+    /// Leaves the left or right part of the trie, depending on the retain mode.
+    ///
+    /// Returns the changes to be applied to in-memory trie and the proof of
+    /// the split operation.
+    pub fn retain_split_shard(
```

Comment:
After analyzing the code and comparing it with what we have in `MemTrieUpdate` for inserts and deletes, the general pattern goes something like this
- We create a new `MemTrieUpdate`
- We call `insert` and `delete` functions on the update to insert or remove trie key/values
- Once we are done with everything, we call `to_mem_trie_changes_only` or `to_trie_changes` that returns us the `TrieChanges`

The interface of the insert and delete functions doesn't return us anything and it's only the final call to `to_trie_changes` that gives us the `TrieChanges` or `MemTrieChanges` object.

I think it should be possible for us to have the `retain_split_shard` function similar to `insert` and `delete` where it doesn't return us anything. If we manage to have the logic of retain to update `tracked_trie_changes` and `updated_nodes`, we should be able to just reuse the logic from `to_trie_changes`.

Thoughts?